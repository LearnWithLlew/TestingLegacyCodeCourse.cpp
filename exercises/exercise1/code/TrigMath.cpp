#include "TrigMath.h"

double TrigMath::SinRaw(int idx) {
    return SIN_TABLE[idx & SIN_MASK];
}

double TrigMath::CosRaw(int idx) {
    return SIN_TABLE[(idx + COS_OFFSET) & SIN_MASK];
}

double TrigMath::Mxatan(double arg) {
    auto argsq = arg * arg;
    auto value = (((p4 * argsq + p3) * argsq + p2) * argsq + p1) * argsq + p0;
    value /= ((((argsq + q4) * argsq + q3) * argsq + q2) * argsq + q1) * argsq + q0;
    return value * arg;
}

double TrigMath::Msatan(double arg) {
    if (arg < sq2m1)
        return Mxatan(arg);
    if (arg > sq2p1)
        return HALF_PI - Mxatan(1 / arg);
    return HALF_PI / 2 + Mxatan((arg - 1) / (arg + 1));
}

void TrigMath::CreateSinTable() {
    for (auto i = 0; i < SIN_SIZE; i++)
        SIN_TABLE[i] = sin(i * TWO_PI / SIN_SIZE);
}

TrigMath::TrigMath() {
    CreateSinTable();
}

double TrigMath::Sin(double angle) {
    return SinRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

int TrigMath::Floor(double a) {
    return (int) a;
}

double TrigMath::Cos(double angle) {
    return CosRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

double TrigMath::Tan(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return SinRaw(idx) / CosRaw(idx);
}

double TrigMath::Sec(double angle) {
    return 1 / Cos(angle);
}

double TrigMath::Cot(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return CosRaw(idx) / SinRaw(idx);
}

double TrigMath::Asin(double value) {
    if (value > 1)
        return NAN;
    if (value < 0)
        return -Asin(-value);
    auto temp = sqrt(1 - value * value);
    if (value > 0.7)
        return HALF_PI - Msatan(temp / value);
    return Msatan(value / temp);
}

double TrigMath::Acos(double value) {
    if (value > 1 || value < -1)
        return NAN;
    return HALF_PI - Asin(value);
}

double TrigMath::Atan(double value) {
    if (value > 0)
        return Msatan(value);
    return -Msatan(-value);
}

double TrigMath::Atan2(double y, double x) {
    if (y + x == y)
        return y >= 0 ? HALF_PI : -HALF_PI;
    y = Atan(y / x);
    if (x < 0) {
        if (y <= 0) {
            return y + PI;
        } else {
            return y - PI;
        }
    }
    return y;
}

double TrigMath::Acsc(double value) {
    if (value == 0)
        return NAN;
    return Asin(1 / value);
}

double TrigMath::Asec(double value) {
    if (value == 0)
        return NAN;
    return Acos(1 / value);
}

double TrigMath::Acot(double value) {
    if (value == 0)
        return NAN;
    if (value > 0)
        return Atan(1 / value);
    return Atan(1 / value) + PI;
}

