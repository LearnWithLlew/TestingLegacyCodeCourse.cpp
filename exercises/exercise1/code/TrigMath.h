//
// Created by LLEWELLYN FALCO on 12/10/17.
//

#ifndef CATCHPLAYGROUND_TRIGMATH_H
#define CATCHPLAYGROUND_TRIGMATH_H

#define _USE_MATH_DEFINES
#include <math.h>
#include <cmath>

class TrigMath {
public:
    const double PI = M_PI;
    const double SQUARED_PI = PI * PI;
    const double HALF_PI = PI / 2;
    const double QUARTER_PI = HALF_PI / 2;
    const double TWO_PI = 2 * PI;
    const double THREE_PI_HALVES = TWO_PI - HALF_PI;
    const double DEG_TO_RAD = PI / 180;
    const double HALF_DEG_TO_RAD = PI / 360;
    const double RAD_TO_DEG = 180 / PI;
    double SQRT_OF_TWO = std::sqrt(2);
    double HALF_SQRT_OF_TWO = SQRT_OF_TWO / 2;
private  :

    const int SIN_SIZE = 100000;
    double SIN_TABLE[100000];
    const int SIN_MASK = SIN_SIZE - 1;
    const double SIN_CONVERSION_FACTOR = SIN_SIZE / TWO_PI;

    const int COS_OFFSET = SIN_SIZE / 4;

//Arc trig
    const double sq2p1 = 2.414213562373095048802;

    const double sq2m1 = 0.414213562373095048802;
    const double p4 = 0.161536412982230228262E2;
    const double p3 = 0.26842548195503973794141E3;
    const double p2 = 0.11530293515404850115428136E4;
    const double p1 = 0.178040631643319697105464587E4;
    const double p0 = 0.89678597403663861959987488E3;
    const double q4 = 0.5895697050844462222791E2;
    const double q3 = 0.536265374031215315104235E3;
    const double q2 = 0.16667838148816337184521798E4;
    const double q1 = 0.207933497444540981287275926E4;
    const double q0 = 0.89678597403663861962481162E3;

    double sinRaw(int idx);

    double cosRaw(int idx);

    double mxatan(double arg);

    double msatan(double arg);

    void createSinTable();

public:

    TrigMath();


    double sin(double angle);

    int floor(double a);


    double cos(double angle);

    double tan(double angle);

    double csc(double angle) {
        return 1 / sin(angle);
    }

    double sec(double angle);


    double cot(double angle);

    double asin(double value);

    double acos(double value);

    double atan(double value);

    double atan2(double y, double x);

    double acsc(double value);

    double asec(double value);

    double acot(double value);


};

#endif //CATCHPLAYGROUND_TRIGMATH_H
