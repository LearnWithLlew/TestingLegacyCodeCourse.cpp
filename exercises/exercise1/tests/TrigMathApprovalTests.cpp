#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"
#include "TrigMath.h"
#include <iostream>
#include <string>
#include <sstream>

using namespace ApprovalTests;



TEST_CASE("ApprovalTest Sin") {
    TrigMath math;
    auto result = math.Sin(0.0);
    Approvals::verify(result);
}

TEST_CASE("ApprovalTest Cos") {
    TrigMath math;
    auto result = math.Cos(0.0);
    Approvals::verify(result);
}

std::string runEverything(double num)
{
    TrigMath math;
    std::stringstream ss;
    ss << "Cos: " << math.Cos(num) << '\n';
    ss << "Sin: " << math.Sin(num) << '\n';
    ss << "Asin: " << math.Asin(num) << '\n';
    ss << "Acos: " << math.Acos(num) << '\n';
    ss << "Acot: " << math.Acot(num) << '\n';
    ss << "Acsc: " << math.Acsc(num) << '\n';
    ss << "Asec: " << math.Asec(num) << '\n';
    ss << "Atan: " << math.Atan(num) << '\n';
    ss << "Atan2: " << math.Atan2(num,0.4) << '\n';
    ss << "Cot: " << math.Cot(num) << '\n';
    ss << "Csc: " << math.Csc(num) << '\n';
    ss << "Sec: " << math.Sec(num) << '\n';
    ss << "Tan: " << math.Tan(num) << '\n';

    return ss.str();
}

TEST_CASE("ApprovalTest Everything 1.0") 
{
    Approvals::verify(runEverything(1.0));
}

TEST_CASE("ApprovalTest Everything 3.14")
{
    Approvals::verify(runEverything(3.14));
}

TEST_CASE("ApprovalTest All")
{
    std::vector<double> inputs{ 1.0, 3.14 , -0.1, 0, 0.2, 0.8 };
    Approvals::verifyAll("TITLE", inputs, [](auto input, auto& stream) {
        stream << input << " => "
            << runEverything(input);
        });
}