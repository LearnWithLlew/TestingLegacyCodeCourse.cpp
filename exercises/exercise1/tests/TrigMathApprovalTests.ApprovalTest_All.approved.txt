TITLE


1 => Cos: 0.828843
Sin: 0.097045
Asin: 1.57075
Acos: 0
Acot: 0.785375
Acsc: 1.57075
Asec: 0
Atan: 0.785375
Atan2: 1.19024
Cot: 8.54081
Csc: 10.3045
Sec: 1.2065
Tan: 0.117085

3.14 => Cos: -0.866633
Sin: 0.866976
Asin: -nan(ind)
Acos: -nan(ind)
Acot: 0.308316
Acsc: 0.324116
Asec: 1.24663
Atan: 1.26243
Atan2: 1.44404
Cot: -0.999604
Csc: 1.15343
Sec: -1.15389
Tan: -1.0004

-0.1 => Cos: 0.033231
Sin: -0.0979798
Asin: -0.100167
Acos: 1.67092
Acot: 1.67042
Acsc: nan
Asec: -nan(ind)
Atan: -0.0996687
Atan2: -0.244979
Cot: -0.339161
Csc: -10.2062
Sec: 30.0924
Tan: -2.94845

0 => Cos: 0.00854478
Sin: 0
Asin: 0
Acos: 1.57075
Acot: -nan(ind)
Acsc: -nan(ind)
Asec: -nan(ind)
Atan: 0
Atan2: 0
Cot: inf
Csc: inf
Sec: 117.031
Tan: 0

0.2 => Cos: 0.0977954
Sin: 0.065234
Asin: 0.201358
Acos: 1.36939
Acot: 1.37335
Acsc: -nan(ind)
Asec: -nan(ind)
Atan: 0.197396
Atan2: 0.463624
Cot: 1.49915
Csc: 15.3294
Sec: 10.2254
Tan: 0.667046

0.8 => Cos: 0.867602
Sin: 0.00980132
Asin: 0.927272
Acos: 0.643478
Acot: 0.896032
Acsc: -nan(ind)
Asec: -nan(ind)
Atan: 0.674718
Atan2: 1.10713
Cot: 88.5188
Csc: 102.027
Sec: 1.1526
Tan: 0.011297


